src/main.rs

```diff
@@ -111,10 +112,7 @@ where
 
         // Grow the buffer if necessary, letting Vec decide what growth
         // factor to use
-        if used == buf.len() {
-            buf.resize(buf.len() + 1, 0);
-            buf.resize(buf.capacity(), 0);
-        }
+        buf.resize(used + 1, 0)
```

Comment:
Let me write some demonstration code:

```rust
fn main() {
    let mut foo = vec![0u8; 4096];
    println!("size={} capacity={}", foo.len(), foo.capacity());
    for _ in 0..10 {
        foo.resize(foo.len() + 1, 0);
        //foo.resize(foo.capacity(), 0);
        println!("size={} capacity={}", foo.len(), foo.capacity());
    }
}
```

We can try with the second line (first) and without (second):

```
size=4096 capacity=4096                                                                                                                  
size=8192 capacity=8192                                                                                                                  
size=16384 capacity=16384                                                                                                                
size=32768 capacity=32768                                                                                                                
size=65536 capacity=65536                                                                                                                
size=131072 capacity=131072                                                                                                              
size=262144 capacity=262144                                                                                                              
size=524288 capacity=524288                                                                                                              
size=1048576 capacity=1048576                                                                                                            
size=2097152 capacity=2097152                                                                                                            
size=4194304 capacity=4194304                                                                                                            
```

```
size=4096 capacity=4096
size=4097 capacity=8192
size=4098 capacity=8192
size=4099 capacity=8192
size=4100 capacity=8192
size=4101 capacity=8192
size=4102 capacity=8192
size=4103 capacity=8192
size=4104 capacity=8192
size=4105 capacity=8192
size=4106 capacity=8192size=4096 capacity=4096
size=4097 capacity=8192
size=4098 capacity=8192
size=4099 capacity=8192
size=4100 capacity=8192
size=4101 capacity=8192
size=4102 capacity=8192
size=4103 capacity=8192
size=4104 capacity=8192
size=4105 capacity=8192
size=4106 capacity=8192
```

In the example with the second line, we can observe exponential; the capacity always matches the size, so there is no unused space in the buffer.

By taking away the second line, the buffer grows linearly while leaving almost 50% of the vector's capacity unused.

We can try an example scenario, e.g. given a 8k buffer, how long would it take to fit a line of length 12k? With the second line, the complexity for fitting a larger string is logarithmic; fitting a 12k string would take just a single extra step (one system call).

Without the second line it would take linear time; specifically it would take around 4k system calls to adjust the buffer to the larger string…

The trick the code uses here is that resize() is allowed to grow the capacity of the buffer beyond the requested size…

Of course, I could achieve the behavior above by simply doing `resize(len() * 2, 0)`; I decided not to because there are actually some vector implementations using different growth factors than two for performance reasons (https://stackoverflow.com/questions/5232198/about-vectors-growth), which is why I decided to not hard code a growth factor myself…

That last point is a bit moot though; in practice resizing the buffer should be a very small part of the applications runtime, since we reuse the buffer so many times…as long as the buffer grows exponentially with *some* growth factor, we're good!

Reply:
I removed the if condition simply because `resize()` does it internally already. Because `buf[used..]` uses `len()`, I resize len to capacity immediately before slicing.

Reply:
Ah I see. And you are growing the buffer to capacity during the read step, so it works…
Ok, just saw that…

So basically in this setup buf.len() is the actual length of the content while the capacity is content + free space? Except in the reading portion of the code, where we resize & truncate the buffer to make sure the above assumption holds water in the rest of the code? With used basically being a temporary for the read portion?

Not sure this is helpful; to be honest I find it kind of hard to read…I thought it was quite nice to have one distinct place in the code that handles buffer size was quite nice. I also liked that the meanings of len() and used didn't change…

Reply:
Yeah, `used` is only a temporary. It's more of my personal style to not change the meaning of `len()`. Not much else.